src/NexusMods.DataModel/FileOriginRegistry.cs

```diff
@@ -60,26 +62,56 @@ public async ValueTask<DownloadId> RegisterDownload(AbsolutePath path, AArchiveM
     /// <inheritdoc />
     public async ValueTask<DownloadId> RegisterFolder(AbsolutePath path, AArchiveMetaData metaData, CancellationToken token = default)
     {
+        List<ArchivedFileEntry> filesToBackup = new();
         List<ArchivedFileEntry> files = new();
         List<RelativePath> paths = new();
 
+        // Build a Hash Table of all currently known files. We do this to deduplicate files between downloads.
+        // TODO: This may need some benchmarking, it's unknown how well this scales across huge loadouts.
+        var existingDownloadIds = _dataStore.AllIds(EntityCategory.DownloadMetadata);
+        var knownHashes = new HashSet<ulong>();
+        foreach (var id in existingDownloadIds)
+        {
+            var ent = _dataStore.Get<DownloadAnalysis>(id);
+
+            // shouldn't be null but technically possible in case of concurrent modifications
+            if (ent == null)
+                continue;
+
+            foreach (var content in ent.Contents)
+                knownHashes.Add(content.Hash.Value);
+        }
```

Comment:
This feels like an suboptimal solution. We should have a map of all files somewhere that we can use for lookup.
Now the question is where. 

I would say in the `NxFileStore` is probably the correct place. 
This means that the map needs to be updated when new files are added and in the future when files will be deleted.

Then the implementation of `NxFileStore.HaveFile` can use the map directly, without having to iterate everything for each request, which sounds quite bad.

Does that make sense? Are there issues with the idea or is it something you could do?

Reply:
It is suboptimal, indeed. I didn't feel like introducing a table just for file hashes as that'd probably be overkill for a single use case.

Pulling this data from `DownloadMetadata` table isn't going to be a bottleneck, what is currently the bottleneck is sending `Get` requests for every single item from the `DataStore` as that involves a lot of back and forth (especially as we scale past 1000 items).

Having a `GetAll` API would solve this problem and be performant enough that it's make a negligible difference in the speed of the overall operation.

I just left it this way for the time being as I want to wait till event sourcing is done. EventSourcing will change database backend from SQLite to RocksDB. 

I'm not sure if @halgari plans to change the interface for this, but there's no point of me writing a `GetAll` for Sqlite if I'll just have to write a `GetAll` method for RocksDB in a month.

Just open an issue for this.

Reply:
Wouldn't adding a `HashSet` of file hashes to `NxFileStore` bypass the need to access the data base? 

Then you get the hit on startup or on first use of `NxFileStore.HaveFile()` if you want to make it lazy, but after that you bypass the `DataStore` right?

This would make `HaveFile` fast, which is what we need here, and would enable us to use it more liberally in the future.

If later we can optimize the first hit on the database to get all the hashes after we switch DB, then that is great.

What do you think?

Reply:
Cool, yeah making it lazy should help.